### PR TITLE
Fix metrics and add a few metrics itests

### PR DESCRIPTION
**Pull request by drolando**
Our refactorings keep messing up our metrics. This PR fixes them and adds a few itests to make sure we don't break them again in the future.

The itest coverage is poor (i.e. I don't test what happens in case of cache hits) but I don't have time right now to write more test cases. We can do that as a follow up PR.

### CODE REVIEW

File: itest/data/srv-configs/casper.internal.yaml [hunk 1]
```diff
@@ -18,7 +18,7 @@ http:
 yelp_meteorite:
     etc_path: '/code/itest/data/etc'
     metrics-relay:
-        host: '127.0.0.1'
+        host: '10.5.0.7'
```
Comment (garylin96) [inline itest/data/srv-configs/casper.internal.yaml]: just out of curiosity, what is the rationale behind this change?

File: itest/metrics/listen.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+
+    # Bind the socket to the port
+    server_address = ('0.0.0.0', 1234)
+    sock.bind(server_address)
+
+    with open('/var/log/metrics/metrics.log', 'w') as fp:
+        while True:
+            data = sock.recv(4096)
+            print(data.decode('utf-8'))
```
Comment (garylin96) [inline itest/metrics/listen.py]: Just to be sure, this is intentional, right?

File: itest/test/spectre/metrics_test.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    dims_list = json.loads(dims_str)
+    for el in dims_list:
+        dims_map[el[0]] = el[1]
+
+    return dims_map
+
+
+def _load_metrics(log_file):
+    metrics = []
+    time.sleep(0.5)
```
Comment (garylin96) [inline itest/test/spectre/metrics_test.py]: Why is this necessary?

File: itest/test/spectre/metrics_test.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    # First 2 metrics are `spectre.fetch_body_and_headers` and `spectre.hit_rate`
+    _assert_fetch_hit_rate(metrics[0:2], 'timestamp')
+
+    # Then since it's a miss we have a `spectre.store_body_and_headers`
+    _assert_store_metric(metrics[2], 'timestamp')
+
+    # Finally the `spectre.request_timing`
+    _assert_request_timing_metrics(metrics[3:7], 'timestamp')
+    assert len(metrics) == 7
```
Comment (garylin96) [inline itest/test/spectre/metrics_test.py]: Shouldn't this assertion be before the `_assert...` calls? If metrics contained less than 7 elements, wouldn't metrics[3:7] raise an exception?

File: itest/test/spectre/metrics_test.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    assert metrics[10].dimensions == {
+        'metric_name': 'spectre.bulk_hit_rate',
+        'habitat': 'uswest1a',
+        'service_name': 'spectre',
+        'namespace': 'backend.main',
+        'instance_name': 'itest',
+        'cache_name': 'bulk_requester_default',
+        'cache_status': 'miss',
+    }
+    assert len(metrics) == 11
```
Comment (garylin96) [inline itest/test/spectre/metrics_test.py]: same as above

File: itest/metrics/listen.py (hunk 2 above)
Comment (drolando) [reply itest/metrics/listen.py]: if you look at `itest/docker-compose.yml`, you’ll see I’ve added a new “metrics” entry with static ip `10.5.0.7`

File: itest/test/spectre/spectre_test.py [hunk 6]
```diff
@@ -641,18 +641,18 @@ def test_different_params_same_id(self):
         assert_is_in_spectre_cache("{}?ids={}&data=false".format(base_path, '6'))
 
     def test_with_invalid_id_when_cache_missing_ids_is_true(self):
-        base_path = '/bulk_requester_2'
+        path = '/bulk_requester_2/{ids}/v1?k1=v1'
 
         # 0 < ids < 1000 are valid
-        resp1 = get_through_spectre("{}/{}/v1?k1=v2".format(base_path, '10,5000,11'))
+        resp1 = get_through_spectre(path.format(ids='10,5000,11'))
```
Comment (garylin96) [inline itest/test/spectre/spectre_test.py]: I know the variable name used to be `resp1`, but could we rename all the variables here that end in 1 or 2 to something better?

Comment (garylin96) [inline itest/test/spectre/spectre_test.py]: I know the variable name used to be resp1, but could we rename all the variables here that end in 1 or 2 to something better?

File: itest/test/spectre/metrics_test.py (hunk 5 above)
Comment (garylin96) [inline itest/test/spectre/metrics_test.py]: same as above

File: itest/test/spectre/metrics_test.py (hunk 3 above)
Comment (drolando) [reply itest/test/spectre/metrics_test.py]: Logs and metrics are emitted after the response is sent back to the client. So we need to wait a bit to make sure spectre had time to write them.

We do the same for `syslog` and those tests don't seem flaky

File: itest/test/spectre/metrics_test.py (hunk 4 above)
Comment (garylin96) [inline itest/test/spectre/metrics_test.py]: Shouldn't this assertion be before the _assert... calls? If metrics contained less than 7 elements, wouldn't metrics[3:7] raise an exception?

File: itest/test/spectre/metrics_test.py (hunk 3 above)
Comment (garylin96) [inline itest/test/spectre/metrics_test.py]: Why is this necessary? Maybe a comment here would be useful

Comment (ArnaudBrousseau) [inline itest/test/spectre/metrics_test.py]: Is the test flaky without this, or straight up failing?

File: itest/test/spectre/metrics_test.py (hunk 4 above)
Comment (drolando) [reply itest/test/spectre/metrics_test.py]: These asserts are mostly there to make sure there are no more metrics than what I expect.

The reason why they're not before the other asserts is because the error message don't show you what metrics are actually in the list, so it's very annoying to figure out what's missing. You'd need to comment out this check and then verify which of the other asserts is failing.

I'll add a comment explaining this.

File: itest/test/spectre/metrics_test.py (hunk 3 above)
Comment (drolando) [reply itest/test/spectre/metrics_test.py]: straight up failing

**Comment by garylin96** [comment]
LGTM! +1
